### ISSUE TITLE
Is there a  breaking change to AQL in 3.2 alpha? 

### ISSUE THREAD

**Issue by mikeatm** [arangodb/arangodb#2431]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [ ] 3.1
- [x] 3.2 pre-releases: ___
- [ ] self-compiled devel branch

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [ ] Linux 
  - [ ] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [x] Fedora .rpm
  - [ ] Gentoo
  - [ ] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

### this is an AQL-related issue:
[x] I'm using graph features

I'm issuing AQL via:
- [ ] web interface with this browser:     running on this OS:
- [x] arangosh
- [ ] this Driver:

I've run `db._explain("<my aql query>")` and it didn't shed more light on this.
The AQL query in question is:
I can not see (or dont understand the  difference  between the two db._explains, but here is part)
3.2 ALPHA

```
Traversals on graphs:
 Id   Depth   Vertex collections   Edge collections   Options                                   Filter conditions
  4   1..1                         publishBY         uniqueVertices: none, uniqueEdges: path 
```
Additionally,  there is a new column `Site` under execution plan, all of it taking the value `COOR`

VER : 3.1.2
```
Traversals on graphs:
 Id   Depth   Vertex collections   Edge collections   Options                                   Filter conditions
  4   1..1                         publishBY         uniqueVertices: none, uniqueEdges: path
```
Additionally, in Execution plan Est is 1

The issue can be reproduced using this dataset:

### this is a web interface-related issue:
I'm using the web interface with this browser:     running on this OS:
- [ ] authentication is enabled?
- [ x] using the cluster?
- [ ] _system database?

These are the steps to reproduce:
1) Open arangosh 
2) Create  graph of { publishingHouse --- publishBY--> publication  }
3) Run query from publication to get publisher. 

The following problem occurs: [Screenshot?] 

This is the error i get on  3.2 apha: (arangosh (ArangoDB 3.2.alpha3 [linux] 64bit, using VPack 0.1.30, ICU 58.1, V8 5.0.71.39, OpenSSL 1.0.2k-fips  26 Jan 2017) )

```
science> db._query("FOR publicat  in @publicsVAR FOR publisher  in INBOUND publicat  publishBY   FILTER publisher !=null  RETURN  publisher",{"publicsVAR": [{"_key": "dftplus","_id":"publication/dftplus"}] }).toArray();
JavaScript exception in file '/usr/share/arangodb3/js/client/modules/@arangodb/arangosh.js' at 97,7: ArangoError 1521: Collection not known to Traversal agency please add 'WITH publishingHouse' as the first line in your AQL (while executing)
!      throw error;
!      ^
stacktrace: ArangoError: Collection not known to Traversal agency please add 'WITH publishingHouse' as the first line in your AQL (while executing)
    at Object.exports.checkRequestResult (/usr/share/arangodb3/js/client/modules/@arangodb/arangosh.js:95:21)
    at ArangoStatement.execute (/usr/share/arangodb3/js/client/modules/@arangodb/arango-statement.js:174:12)
    at ArangoDatabase._query (/usr/share/arangodb3/js/client/modules/@arangodb/arango-database.js:909:42)
    at <shell command>:1:4

```

Instead I would be expecting: 
This is the result on 3.1.2 
```
127.0.0.1:8529@science> db._query("FOR publicat  in @publicsVAR FOR publisher  in INBOUND publicat  publishBY   FILTER publisher !=null  RETURN  publisher",{"publicsVAR": [{"_key": "dftplus","_id":"publication/dftplus"}] }).toArray();
[ 
  { 
    "_key" : "powser",
    "_id" : "publishingHouse/powser",
    "_rev" : "_UNRGO9q--_",
    "email" : "two@one.com",
    "name" : "powser"
  }
]
```

Adding the advised command fixes it: 
`db._query(" WITH publishingHouse  FOR publicat  in @publicsVAR FOR publisher  in INBOUND publicat  publishBY   FILTER publisher !=null  RETURN  publisher",{"publicsVAR": [{"_key": "dftplus","_id":"publication/dftplus"}] }).toArray();
`
### this is an installation-related issue:
Describe which steps you carried out, what you expected to happen and what actually happened.

**Comment by jsteemann** [comment]
As far as I can see, the issue also occurs in 3.1.
I have tried the latest 3.1 and 3.1.2. It occurs in both versions when I use the query shown above.
However, and this may have caused the confusion, it will only occur when there is a connecting edge found in the traversal that links to a collection not declared in the query.

The AQL query explicitly uses some collections (e.g. `publishBy`) but some may only be found dynamically during the traversal (e.g. `publishingHouse` or `agency`). Those collections are not mentioned in the query string, so they are only found when an edge is encountered during the traversal that links to them. This is unsupported in the cluster for some technical reasons. In the cluster, the collections the traversal will use need to be known in advance. The fix for this is to declare those collections with the `WITH` statement at the beginning of the query.

This `WITH` declaration is not required in single-server mode. If you ran the query in 3.1.2 in single-server mode, the query should work without the `WITH declaration`. This is also true for 3.2 alpha.
 
If you ran the query in a 3.1.2 cluster, then it should only work there when there are no edges found during the traversal that connect to non-declared collections. The same is true for 3.2.

I think we made the `WITH` declaration mandatory starting with 3.1.0, but not with later releases (i.e. between 3.1.2 and 3.2 alpha).
